File: ScaleformUI_Csharp/LobbyMenu/Panels/PlayerStatsPanel.cs
using CitizenFX.Core;
using ScaleformUI.PauseMenu;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ScaleformUI.LobbyMenu
{
    public class PlayerStatsPanel
    {
        public LobbyItem ParentItem { get; set; }
        private string title;
        private string description;
        private HudColor titleColor = HudColor.HUD_COLOUR_FREEMODE;
        private bool hasPlane = false;
        private bool hasVehicle = false;
        private bool hasBoat = false;
        private bool hasHeli = false;

        public UpperInformation RankInfo { get; set; }
        public string Title
        {
            get => title;
            set
            {
                title = value;
                UpdatePanel();
            }
        }
        public HudColor TitleColor
        {
            get => titleColor;
            set
            {
                titleColor = value;
                UpdatePanel();
            }
        }
        public string Description
        {
            get => description;
            set
            {
                description = value;
                UpdatePanel();
            }
        }
        public bool HasPlane
        {
            get => hasPlane;
            set
            {
                hasPlane = value;
                UpdatePanel();
            }
        }
        public bool HasHeli
        {
            get => hasHeli;
            set
            {
                hasHeli = value;
                UpdatePanel();
            }
        }
        public bool HasBoat
        {
            get => hasBoat;
            set
            {
                hasBoat = value;
                UpdatePanel();
            }
        }
        public bool HasVehicle
        {
            get => hasVehicle;
            set
            {
                hasVehicle = value;
                UpdatePanel();
            }
        }
        public List<PlayerStatsPanelStatItem> Items { get; private set; }
        public PlayerStatsPanel(string title, HudColor titleColor)
        {
            this.title = title;
            this.titleColor = titleColor;
            this.description = string.Empty;
            RankInfo = new(this);
            Items = new();
        }

        public void AddStat(PlayerStatsPanelStatItem item)
        {
            item.Parent = this;
            item.idx = Items.Count;
            Items.Add(item);
            UpdatePanel();
        }

        public void UpdatePanel(bool _override = false)
        {
            if ((ParentItem!= null && ParentItem.ParentColumn != null && ParentItem.ParentColumn.Parent != null && ParentItem.ParentColumn.Parent.Visible) || _override)
            {
                var idx = ParentItem.ParentColumn.Items.IndexOf(ParentItem);
                if (ParentItem.ParentColumn.Parent is MainView lobby)
                {
                    lobby._pause._lobby.CallFunction("SET_PLAYER_ITEM_PANEL", idx, 0, ParentItem.ClonePed != null, Title, Description, (int)TitleColor, RankInfo.RankLevel, HasPlane, HasHeli, HasBoat, HasVehicle, 0, RankInfo.LowLabel, 0, 0, RankInfo.MidLabel, 0, 0, RankInfo.UpLabel, 0, 0);
                    if (!string.IsNullOrWhiteSpace(Description))
                        lobby._pause._lobby.CallFunction("SET_PLAYER_ITEM_PANEL_DESCRIPTION", idx, Description, 0, "", ParentItem.ClonePed != null);
                    foreach (var stat in Items)
                        lobby._pause._lobby.CallFunction("SET_PLAYER_ITEM_PANEL_STAT", idx, stat.idx, 0, stat.Label, stat.Description, stat.Value);
                }
                else if (ParentItem.ParentColumn.Parent is TabView pause)
                {
                    pause._pause._pause.CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL", ParentItem.ParentColumn.ParentTab, idx, 0,ParentItem.ClonePed != null,  Title, Description, (int)TitleColor, RankInfo.RankLevel, HasPlane, HasHeli, HasBoat, HasVehicle, 0, RankInfo.LowLabel, 0, 0, RankInfo.MidLabel, 0, 0, RankInfo.UpLabel, 0, 0);
                    if (!string.IsNullOrWhiteSpace(Description))
                        pause._pause._pause.CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL_DESCRIPTION", ParentItem.ParentColumn.ParentTab, idx, Description, 0, "", ParentItem.ClonePed != null);
                    foreach (var stat in Items)
                        pause._pause._pause.CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL_STAT", ParentItem.ParentColumn.ParentTab, idx, stat.idx, 0, stat.Label, stat.Description, stat.Value);
                }
            }
        }
    }

    public sealed class UpperInformation
    {
        internal PlayerStatsPanel _parent;
        private int rankLevel = 0;
        private string upLabel = "";
        private string lowLabel = "";
        private string midLabel = "";

        public int RankLevel
        {
            get => rankLevel;
            set
            {
                rankLevel = value;
                _parent.UpdatePanel();
            }
        }
        public string UpLabel
        {
            get => upLabel;
            set
            {
                upLabel = value;
                _parent.UpdatePanel();
            }
        }
        public string MidLabel
        {
            get => midLabel;
            set
            {
                midLabel = value;
                _parent.UpdatePanel();
            }
        }
        public string LowLabel
        {
            get => lowLabel;
            set
            {
                lowLabel = value;
                _parent.UpdatePanel();
            }
        }
        public UpperInformation(PlayerStatsPanel parent)
        {
            _parent = parent;
        }
    }

    public sealed class PlayerStatsPanelStatItem
    {

        internal int idx;
        private int _value;
        private string description;
        private string label;
        public PlayerStatsPanel Parent { get; internal set; }
        public string Label
        {
            get => label;
            set
            {
                label = value;
                Parent.UpdatePanel();
            }
        }
        public string Description
        {
            get => description;
            set
            {
                description = value;
                Parent.UpdatePanel();
            }
        }
        public int Value
        {
            get => _value;
            set
            {
                _value = value;
                Parent.UpdatePanel();
            }
        }

        public PlayerStatsPanelStatItem(string label, string description, int value)
        {
            this.label = label;
            this.description = description;
            this._value = value;
        }
    }
}

File: ScaleformUI_Lua/src/LobbyMenu/Items/ItemPanels/PlayerStatsPanel.lua
PlayerStatsPanel = setmetatable({}, PlayerStatsPanel)
PlayerStatsPanel.__index = PlayerStatsPanel
PlayerStatsPanel.__call = function()
    return "ItemPanel", "PlayerStatsPanel"
end

---@class PlayerStatsPanel
---@field private _title string
---@field private _description string
---@field private _titleColor number
---@field private _hasPlane boolean
---@field private _hasVehicle boolean
---@field private _hasBoat boolean
---@field private _hasHeli boolean
---@field public ParentItem FriendItem
---@field public RankInfo UpperInformation
---@field public Items PlayerStatsPanelStatItem[]
---@field public UpdatePanel fun(override: boolean|nil)
---@field public OnItemChanged fun(item: PlayerStatsPanelStatItem)
---@field public OnItemActivated fun(item: PlayerStatsPanelStatItem)
---@field public Description fun(desc: string|nil): string
---@field public HasPlane fun(bool: boolean|nil): boolean
---@field public HasHeli fun(bool: boolean|nil): boolean
---@field public HasBoat fun(bool: boolean|nil): boolean
---@field public HasVehicle fun(bool: boolean|nil): boolean
---@field public AddStat fun(statItem: PlayerStatsPanelStatItem)

---Creates a new PlayerStatsPanel.
---@param title string
---@param titleColor number
---@return PlayerStatsPanel
function PlayerStatsPanel.New(title, titleColor)
    local _data = {
        ParentItem = nil,
        _title = title or "",
        _description = "",
        _titleColor = titleColor or 116,
        _hasPlane = false,
        _hasVehicle = false,
        _hasBoat = false,
        _hasHeli = false,
        RankInfo = nil,
        Items = {}
    }
    local retVal = setmetatable(_data, PlayerStatsPanel)
    retVal.RankInfo = UpperInformation.New(retVal)
    return retVal
end

---Sets the title of the panel if supplied else it will return the current title.
---@param label string|nil
---@return string
function PlayerStatsPanel:Title(label)
    if label ~= nil then
        self._title = label
        self:UpdatePanel()
    end
    return self._title
end

---Sets the title color of the panel if supplied else it will return the current color.
---@param color number|nil
---@return number
function PlayerStatsPanel:TitleColor(color)
    if color ~= nil then
        self._titleColor = color
        self:UpdatePanel()
    end
    return self._titleColor
end

---Sets the description of the panel if supplied else it will return the current description.
---@param label string|nil
---@return string
function PlayerStatsPanel:Description(label)
    if label ~= nil then
        self._description = label
        self:UpdatePanel()
    end
    return self._description
end

---Sets whether the player has a plane or not, if parameter is nill, it will return the current value.
---@param bool boolean|nil
---@return boolean
function PlayerStatsPanel:HasPlane(bool)
    if bool ~= nil then
        self._hasPlane = bool
        self:UpdatePanel()
    end
    return self._hasPlane
end

---Sets whether the player has a helicopter or not, if parameter is nill, it will return the current value.
---@param bool boolean|nil
---@return boolean
function PlayerStatsPanel:HasHeli(bool)
    if bool ~= nil then
        self._hasHeli = bool
        self:UpdatePanel()
    end
    return self._hasHeli
end

---Sets whether the player has a boat or not, if parameter is nill, it will return the current value.
---@param bool boolean|nil
---@return boolean
function PlayerStatsPanel:HasBoat(bool)
    if bool ~= nil then
        self._hasBoat = bool
        self:UpdatePanel()
    end
    return self._hasBoat
end

---Sets whether the player has a vehicle or not, if parameter is nill, it will return the current value.
---@param bool boolean|nil
---@return boolean
function PlayerStatsPanel:HasVehicle(bool)
    if bool ~= nil then
        self._hasVehicle = bool
        self:UpdatePanel()
    end
    return self._hasVehicle
end

---Adds a new stat item to the panel.
---@param statItem PlayerStatsPanelStatItem
function PlayerStatsPanel:AddStat(statItem)
    statItem.Parent = self
    statItem.idx = #self.Items
    self.Items[#self.Items + 1] = statItem
    self:UpdatePanel()
end

---Triggers the panel to update.
---@param override boolean|nil If true, the panel will update regardless of the parent's visibility.
function PlayerStatsPanel:UpdatePanel(override)
    if override == nil then override = false end
    if ((self.ParentItem ~= nil and self.ParentItem.ParentColumn ~= nil and self.ParentItem.ParentColumn.Parent ~= nil and self.ParentItem.ParentColumn.Parent:Visible()) or override) then
        local idx = IndexOf(self.ParentItem.ParentColumn.Items, self.ParentItem) - 1
        local pSubT = self.ParentItem.ParentColumn.Parent()
        if pSubT == "LobbyMenu" then
            ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_PLAYER_ITEM_PANEL", false, idx, 0,
                self.ParentItem.ClonePed ~= nil and self.ParentItem.ClonePed ~= 0, self:Title(), self:Description(),
                self:TitleColor(), self.RankInfo:RankLevel(), self:HasPlane(), self:HasHeli(), self:HasBoat(),
                self:HasVehicle(), 0, self.RankInfo:LowLabel(), 0, 0, self.RankInfo:MidLabel(), 0, 0,
                self.RankInfo:UpLabel(),
                0, 0)
            if not self:Description():IsNullOrEmpty() then
                ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_PLAYER_ITEM_PANEL_DESCRIPTION", false, idx,
                    self:Description(), 0, "", self.ParentItem.ClonePed ~= nil and self.ParentItem.ClonePed ~= 0)
            end
            for k, stat in pairs(self.Items) do
                ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_PLAYER_ITEM_PANEL_STAT", false, idx, stat.idx,
                    0, stat:Label(), stat:Description(), stat:Value())
            end
        elseif pSubT == "PauseMenu" then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL", false,
                self.ParentItem.ParentColumn.ParentTab, idx, 0,
                self.ParentItem.ClonePed ~= nil and self.ParentItem.ClonePed ~= 0, self:Title(), self:Description(),
                self:TitleColor(), self.RankInfo:RankLevel(), self:HasPlane(), self:HasHeli(), self:HasBoat(),
                self:HasVehicle(), 0, self.RankInfo:LowLabel(), 0, 0, self.RankInfo:MidLabel(), 0, 0,
                self.RankInfo:UpLabel(),
                0, 0)
            if not self:Description():IsNullOrEmpty() then
                ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL_DESCRIPTION",
                    false, self.ParentItem.ParentColumn.ParentTab, idx, self:Description(), 0, "",
                    self.ParentItem.ClonePed ~= nil and self.ParentItem.ClonePed ~= 0)
            end
            for k, stat in pairs(self.Items) do
                ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_PLAYER_ITEM_PANEL_STAT", false,
                    self.ParentItem.ParentColumn.ParentTab, idx, stat.idx, 0, stat:Label(), stat:Description(),
                    stat:Value())
            end
        end
    end
end
